Add user-friendly high-level ES|QL unmapped fields overview (#151419)
elastic/elasticsearch: docs/reference/query-languages/esql/esql-unmapped-fields.md

---
applies_to:
  stack: preview 9.3.0
  serverless: preview
navigation_title: "Unmapped fields"
description: How ES|QL queries fields that aren't in the index mapping, using the SET unmapped_fields directive.
---

# Query unmapped fields without reindexing [esql-unmapped-fields]

{{esql}} can query fields that are not defined in your index mapping, with no reindex and no change to the mapping. The [`SET unmapped_fields`](commands/set.md#esql-unmapped_fields) directive controls how each query handles them.

> An unmapped field is a field that a query references but that the mapping does not define. By default, {{esql}} returns an error for it.

Reviewer: nit: there is some nuance that is glossed over here. It may be ok for a simplified intro, though.

An unmapped field is a field in a specific index that the mapping doesn't define, but is present in the indexed documents. As long as it is in `_source`, we make it possible to query it. When you reference a field in a query that is unmapped in all indices, then we error out - otherwise, we just fill it with nulls.

So, maybe this is more precise:

```suggestion
An unmapped field is a field in indexed documents that the index' mapping does not define. By default, {{esql}} treats such fields as `null` and returns an error if a referenced field is not mapped in any index.
```

Author: 100% 👍

> Without it, you need to reindex your data so the field is mapped before you can query it. On a large dataset, that reindex can take hours.

Reviewer 2: ```suggestion
Without this capability, the usual fix is to add the field to your mapping and reindex your data before you can query it. On a large dataset, that reindex can take hours.
```

## Use cases [esql-unmapped-fields-use-cases]

Querying unmapped fields helps in several common situations where the mapping does not yet match the data you want to explore:

- **Explore new fields before mapping them**: query the values of a field as soon as it appears, whether it arrives unexpectedly from an integration or an upstream change, then decide later whether the performance gains of formal mapping justify a reindex.
- **Read the real values of a partially mapped field**: when a field is mapped in some indices but not others, get its values everywhere in a single query, including the indices where it is unmapped.

> - **Keep result columns consistent**: fill a missing field with `null` across a wildcard pattern, a [view](esql-views.md), or a [subquery](esql-subquery.md), so that the set of columns stays stable instead of raising an error.

Reviewer: I think this is a hallucination, or at least I don't grok what this sentence means.

Partially mapped fields are already filled with nulls. Result columns are made consistent automatically when using subqueries and views, that's just ESQL's standard union semantics in `FROM idx, view, (FROM subquery-idx), ...`.

The `nullify` setting fills columns with nulls that are otherwise non-existent and would raise a verification exception. The `load` setting tries to load from `_source` for such fields. The use case to mention here is that you can make queries re-usable across different index patterns, and make queries resilient to mapping changes in datastream rollovers that change the mapping. You may have a prepared query like `FROM logs-* | EVAL foo = bar * 2` and you copy-paste it from your friend; the `bar` field might be pretty standard to have, but maybe your dataset doesn't have it. And boom! Default ESQL will bail, even though the rest of the query might still be useful to run. `nullify` and `load` give you tools to run the query, anyway, either ignoring unmapped data or trying to load it.

:::{tip}

> Some data lives inside JSON strings or [`flattened`](/reference/elasticsearch/mapping-reference/flattened.md) fields, where loading the whole field is not enough. To pull a specific value out of that nested structure, use the [`JSON_EXTRACT`](functions-operators/string-functions/json_extract.md) function.

Reviewer: I don't know that JSON_EXTRACT will work with flattened fields; this is likely a misnomer. @nik9000 will know this better.

The tip we can give is: Vanilla ESQL let's you extract stuff from `_source` using `JSON_EXTRACT`. This way, you can extract _some_ unmapped fields if you wish, but you don't get automagicks without `load`.

Author: could wait to publish until https://github.com/elastic/elasticsearch/pull/151841 gets in?

Reviewer: That other PR is using FIELD_EXTRACT which is specifically working with flattened fields and has slightly different semantics to JSON_EXTRACT AFAIR.

Both things are probably worth mentioning:
- JSON_EXTRACT lets' you extract all you want from `_source` (already)
- FIELD_EXTRACT will let you extract subfields of flattened fields, specifically (once it's merged, but it's planned to be de-snapshotted any moment now)

Waiting for https://github.com/elastic/elasticsearch/pull/151841 and getting a second pair of eyes from @nik9000 or @leontyevdv on this specific paragraph is likely a good idea.

> To extract a value from a JSON string or directly from `_source`, use [`JSON_EXTRACT`](functions-operators/string-functions/json_extract.md). To extract a subfield from a [`flattened`](/reference/elasticsearch/mapping-reference/flattened.md) field, use `FIELD_EXTRACT` {applies_to}`stack: preview 9.5.0` {applies_to}`serverless: preview`.

Author: @nik9000 @leontyevdv lemme know if this looks good to you :)

Reviewer 3: Hi @leemthompo! Yes, everything is fine. I just noticed a few things:
- We use 9.3, 9.4, etc. everywhere in this doc and 9.5.0 for the FIELD_EXTRACT (technically, it's correct but not consistent)
- Add a link once your other PR has been merged.
- JSON_EXTRACT has no `applies_to` badge. The page header contains preview 9.3.0 but JSON_EXTRACT only exists from 9.4 (preview) / 9.5 (GA)

:::

## How {{esql}} handles unmapped fields [esql-unmapped-fields-how-it-works]

The `unmapped_fields` setting accepts three values, which range from strict to permissive. These values also control how {{esql}} resolves a partially unmapped field, one that exists in the mapping of some indices but not others, across a multi-index query.

| Value | What it does | When to use it |
| --- | --- | --- |

> | `DEFAULT` | The query fails when it references an unmapped field. This is the default behavior. | You want strict schema enforcement and prefer an error over silently missing data. |

Reviewer: We could also mention what happens to partially mapped fields. (nullified in DEFAULT and NULLIFY, loaded from source for LOAD)

> | `NULLIFY` | Unmapped fields come back as columns of `null` values. | You query several indices with a wildcard and a field is missing from some of them, or you work with views and subqueries that need a stable set of columns. |

Reviewer: I still don't know what "stable set of columns" is supposed to mean, and how this relates to views and subqueries.

Author: ❌ 🖍️

> | `LOAD` {applies_to}`stack: preview 9.4` | {{esql}} loads unmapped fields from the stored [`_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md) at search time and treats them as `keyword`. Fields absent from `_source` come back as `null`. | You need the real values of a field that was never mapped, so that you can filter or aggregate on it. |

Reviewer: Yeah, this is fairly precise and concise, nice.

> For the full syntax and the accepted values, refer to the [`SET unmapped_fields`](commands/set.md#esql-unmapped_fields) reference.

Reviewer: nit: all accepted values are already mentioned above.

```suggestion
For the full syntax, refer to the [`SET unmapped_fields`](commands/set.md#esql-unmapped_fields) reference.
```

:::{tip}
:applies_to: stack: preview 9.4

> Reading the real values of an unmapped field with `LOAD` is newer than the other behaviors. Earlier versions support `NULLIFY` only, which fills unmapped fields with `null`.

Reviewer 2: ```suggestion
Reading the real values of an unmapped field with `LOAD` is newer than the other behaviors. Earlier versions can nullify unmapped fields (`NULLIFY`) but can't load their real values.
```

:::

## Limitations [esql-unmapped-fields-limitations]

`LOAD` does not support every command, function, and field type. For the current restrictions, refer to the [`SET unmapped_fields`](commands/set.md#esql-unmapped_fields) reference.

> One restriction worth planning around is performance: reading from the stored [`_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md) is slower than querying a mapped field, because there is no inverted index to accelerate filters. The trade-off is that the data is available in seconds instead of after a full reindex.

Reviewer 4: I would replace "inverted index" by something more generic, since this also applies to other kinds of value storage such as doc values, time series, etc which provide fast access but technically aren't inverted indexes. Something like "there is no data preparation to accelerate filters" or "since the value hasn't been stored in a fast-access structure"?

Author: > I would replace "inverted index" by something more generic

💯

> One restriction worth planning around is performance: reading from the stored [`_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md) is slower than querying a mapped field, because the values aren't stored in a data structure optimized for fast access. The trade-off is that the data is available in seconds instead of after a full reindex.

Reviewer: nit: it's not always seconds, it depends on the query. For very tiny data sets, there may even be no noticeable difference. Maybe this is better:

```suggestion
One restriction worth planning around is performance: reading from the stored [`_source`](/reference/elasticsearch/mapping-reference/mapping-source-field.md) is slower than querying a mapped field, because the values aren't stored in a data structure optimized for fast access. The trade-off is that the data is available immediately without a full reindex.
```

## Unmapped fields and runtime fields [esql-unmapped-fields-vs-runtime-fields]

> Unmapped fields differ from runtime fields. A runtime field is a computed field defined in the index mapping, and {{esql}} treats it like any other mapped field. You cannot define new runtime fields at search time in {{esql}}. To create computed columns instead, use the [`EVAL`](commands/eval.md) command. To learn more, refer to [runtime fields](docs-content://manage-data/data-store/mapping/runtime-fields.md).

Reviewer: nit: the way we speak about runtime fields in `_search`, it doesn't even have to be in the mapping, it can be defined in the request.
```suggestion
Unmapped fields differ from runtime fields. A runtime field is a computed field that can be defined in the index mapping, while an unmapped field is not in the mapping at all but can be present in documents. If a runtime field is part of the index mapping, {{esql}} treats it like any other mapped field. In {{esql}}, you create computed columns with the [`EVAL`](commands/eval.md) command. To learn more, refer to [runtime fields](docs-content://manage-data/data-store/mapping/runtime-fields.md).
```

## Related resources [esql-unmapped-fields-related]

To go deeper on unmapped fields and related capabilities, refer to these pages:

- [`SET unmapped_fields`](commands/set.md#esql-unmapped_fields): the directive syntax, accepted values, and `LOAD` limitations.

> - [`JSON_EXTRACT`](functions-operators/string-functions/json_extract.md): extract values from JSON strings and `flattened` fields.

Reviewer: The flattened field type is still unsupported.
```suggestion
- [`JSON_EXTRACT`](functions-operators/string-functions/json_extract.md): extract values from JSON strings and `_source`.
```

Author: I guess I could wait until https://github.com/elastic/elasticsearch/pull/151841 gets in? @nik9000 @leontyevdv WDYT?

Reviewer 3: Hi @leemthompo! Yes, I think we'll merge it soon. 

A comment about the content:
The 'json_extract()' function cannot accept a flattened type as input. There is now a specific function for that called 'field_extract()'.
